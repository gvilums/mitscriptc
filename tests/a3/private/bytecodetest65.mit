x = 5;
foo = fun(a, s, df) {
    print(a);
    print(s);
    print(df);
    print(x);
    x = 3;
    print(x);
    foo2 = fun() {
        print(x);
        foo3 = fun(x) {
            print(x);
            foo4 = fun() {
                print(x);
                x = 9;
                while (false) {
                    global x;
                }
                foo5 = fun() {
                    print(x + 1);
                    if (false) {
                        foo6 = fun() {};
                    }
                    return foo6;
                };
                return foo5;
            };
            return foo4;
        };
        return foo3;
    };
    x = 4;
    return foo2;
};
a = foo;
a = a(1, 2, 3);
a = a();
a = a(a);
a();
a = a();
print(a());
print(a == foo);
x = -1;
a();

bar = fun() {
    print(x);
    return None;
    x = 3;
};
bar();
bar = fun() {
    print(x);
    if (false) {
        x = 3;
    }
};
bar();
bar();

print("frame");
oldfun = None;
fgen = fun(x) {
    global oldfun;
    newfun = fun() {
        print(x);
    };
    newfun();
    print(newfun == oldfun);
    oldfun = newfun;
    return newfun;
};
a = fgen(0);
b = fgen(0);
fgen(0);
c = fgen(1);
a();
b();
c();

print("while");

oldfun = None;
i = 0;
rec = {};
while (i < 3) {
    newfun = fun() {
        print(i);
    };
    newfun();
    //print(newfun == oldfun);
    oldfun = newfun;
    rec[i] = newfun;
    i = i + 1;
}
rec[0]();
rec[1]();
rec[2]();
