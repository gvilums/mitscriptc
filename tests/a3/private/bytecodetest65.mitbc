function
{
	functions =
	[
		function
		{
			functions = [],
			constants = [],
			parameter_count = 1,
			local_vars = [],
			local_ref_vars = [],
			free_vars = [],
			names = [],
			instructions = 
			[
			]
		},
		function
		{
			functions = [],
			constants = [],
			parameter_count = 0,
			local_vars = [],
			local_ref_vars = [],
			free_vars = [],
			names = [],
			instructions = 
			[
			]
		},
		function
		{
			functions = [],
			constants = [],
			parameter_count = 1,
			local_vars = [],
			local_ref_vars = [],
			free_vars = [],
			names = [],
			instructions = 
			[
			]
		},
		function
		{
			functions =
			[
				function
				{
					functions =
					[
						function
						{
							functions =
							[
								function
								{
									functions =
									[
										function
										{
											functions =
											[
												function
												{
													functions = [],
													constants = [None],
													parameter_count = 0,
													local_vars = [],
													local_ref_vars = [],
													free_vars = [],
													names = [],
													instructions = 
													[
														load_const	0
														return
													]
												}
											],
											constants = [None, 1, false],
											parameter_count = 0,
											local_vars = [foo6],
											local_ref_vars = [],
											free_vars = [],
											names = [print, x],
											instructions = 
											[
												load_const	0
												store_local	0
												load_global	0
												load_global	1
												load_const	1
												add
												call	1
												pop
												load_const	2
												not
												if	4
												load_func	0
												alloc_closure	0
												store_local	0
												load_local	0
												return
											]
										}
									],
									constants = [None, 9, false],
									parameter_count = 0,
									local_vars = [foo5],
									local_ref_vars = [],
									free_vars = [],
									names = [x, print],
									instructions = 
									[
										load_const	0
										store_local	0
										load_global	1
										load_global	0
										call	1
										pop
										load_const	1
										store_global	0
										load_const	2
										not
										if	2
										goto	-3
										load_func	0
										alloc_closure	0
										store_local	0
										load_local	0
										return
									]
								}
							],
							constants = [None],
							parameter_count = 1,
							local_vars = [x, foo4],
							local_ref_vars = [],
							free_vars = [],
							names = [print],
							instructions = 
							[
								load_const	0
								store_local	1
								load_global	0
								load_local	0
								call	1
								pop
								load_func	0
								alloc_closure	0
								store_local	1
								load_local	1
								return
							]
						}
					],
					constants = [None],
					parameter_count = 0,
					local_vars = [foo3],
					local_ref_vars = [],
					free_vars = [x],
					names = [print],
					instructions = 
					[
						load_const	0
						store_local	0
						load_global	0
						push_ref	0
						load_ref
						call	1
						pop
						load_func	0
						alloc_closure	0
						store_local	0
						load_local	0
						return
					]
				}
			],
			constants = [None, 3, 4],
			parameter_count = 3,
			local_vars = [a, s, df, x, foo2],
			local_ref_vars = [x],
			free_vars = [],
			names = [print],
			instructions = 
			[
				load_const	0
				store_local	3
				load_const	0
				store_local	4
				load_global	0
				load_local	0
				call	1
				pop
				load_global	0
				load_local	1
				call	1
				pop
				load_global	0
				load_local	2
				call	1
				pop
				load_global	0
				load_local	3
				call	1
				pop
				load_const	1
				store_local	3
				load_global	0
				load_local	3
				call	1
				pop
				load_func	0
				push_ref	0
				alloc_closure	1
				store_local	4
				load_const	2
				store_local	3
				load_local	4
				return
			]
		},
		function
		{
			functions = [],
			constants = [None, 3],
			parameter_count = 0,
			local_vars = [x],
			local_ref_vars = [],
			free_vars = [],
			names = [print],
			instructions = 
			[
				load_const	0
				store_local	0
				load_global	0
				load_local	0
				call	1
				pop
				load_const	0
				return
				load_const	1
				store_local	0
			]
		},
		function
		{
			functions = [],
			constants = [None, false, 3],
			parameter_count = 0,
			local_vars = [x],
			local_ref_vars = [],
			free_vars = [],
			names = [print],
			instructions = 
			[
				load_const	0
				store_local	0
				load_global	0
				load_local	0
				call	1
				pop
				load_const	1
				not
				if	3
				load_const	2
				store_local	0
				load_const	0
				return
			]
		},
		function
		{
			functions =
			[
				function
				{
					functions = [],
					constants = [None],
					parameter_count = 0,
					local_vars = [],
					local_ref_vars = [],
					free_vars = [x],
					names = [print],
					instructions = 
					[
						load_global	0
						push_ref	0
						load_ref
						call	1
						pop
						load_const	0
						return
					]
				}
			],
			constants = [None],
			parameter_count = 1,
			local_vars = [x, newfun],
			local_ref_vars = [x],
			free_vars = [],
			names = [oldfun, print],
			instructions = 
			[
				load_const	0
				store_local	1
				load_func	0
				push_ref	0
				alloc_closure	1
				store_local	1
				load_local	1
				call	0
				pop
				load_global	1
				load_local	1
				load_global	0
				eq
				call	1
				pop
				load_local	1
				store_global	0
				load_local	1
				return
			]
		},
		function
		{
			functions = [],
			constants = [None],
			parameter_count = 0,
			local_vars = [],
			local_ref_vars = [],
			free_vars = [],
			names = [print, i],
			instructions = 
			[
				load_global	0
				load_global	1
				call	1
				pop
				load_const	0
				return
			]
		}
	],
	constants = [5, 1, 2, 3, "frame", None, 0, "while"],
	parameter_count = 0,
	local_vars = [],
	local_ref_vars = [],
	free_vars = [],
	names = [x, foo, a, bar, oldfun, fgen, b, c, i, rec, newfun, print, input, intcast],
	instructions = 
	[
		load_func	0
		alloc_closure	0
		store_global	11
		load_func	1
		alloc_closure	0
		store_global	12
		load_func	2
		alloc_closure	0
		store_global	13
		load_const	0
		store_global	0
		load_func	3
		alloc_closure	0
		store_global	1
		load_global	1
		store_global	2
		load_global	2
		load_const	1
		load_const	2
		load_const	3
		call	3
		store_global	2
		load_global	2
		call	0
		store_global	2
		load_global	2
		load_global	2
		call	1
		store_global	2
		load_global	2
		call	0
		pop
		load_global	2
		call	0
		store_global	2
		load_global	11
		load_global	2
		call	0
		call	1
		pop
		load_global	11
		load_global	2
		load_global	1
		eq
		call	1
		pop
		load_const	1
		neg
		store_global	0
		load_global	2
		call	0
		pop
		load_func	4
		alloc_closure	0
		store_global	3
		load_global	3
		call	0
		pop
		load_func	5
		alloc_closure	0
		store_global	3
		load_global	3
		call	0
		pop
		load_global	3
		call	0
		pop
		load_global	11
		load_const	4
		call	1
		pop
		load_const	5
		store_global	4
		load_func	6
		alloc_closure	0
		store_global	5
		load_global	5
		load_const	6
		call	1
		store_global	2
		load_global	5
		load_const	6
		call	1
		store_global	6
		load_global	5
		load_const	6
		call	1
		pop
		load_global	5
		load_const	1
		call	1
		store_global	7
		load_global	2
		call	0
		pop
		load_global	6
		call	0
		pop
		load_global	7
		call	0
		pop
		load_global	11
		load_const	7
		call	1
		pop
		load_const	5
		store_global	4
		load_const	6
		store_global	8
		alloc_record
		store_global	9
		load_global	8
		load_const	3
		swap
		gt
		not
		if	18
		load_func	7
		alloc_closure	0
		store_global	10
		load_global	10
		call	0
		pop
		load_global	10
		store_global	4
		load_global	9
		load_global	8
		load_global	10
		index_store
		load_global	8
		load_const	1
		add
		store_global	8
		goto	-22
		load_global	9
		load_const	6
		index_load
		call	0
		pop
		load_global	9
		load_const	1
		index_load
		call	0
		pop
		load_global	9
		load_const	2
		index_load
		call	0
		pop
	]
}