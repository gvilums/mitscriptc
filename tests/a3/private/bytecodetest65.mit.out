1
2
3
None
3
4
FUNCTION
5
9
10
None
false
0
None
None
None
frame
0
false
0
false
0
false
1
false
0
0
1
while
0
1
2
3
3
3
